include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -961,135 +970,449 @@ void reduceSumScalingVector(common::Pipeline<comm::Communicator>& col_task_chain
-          const SizeType i_subm_i2_arr = dist.globalTileFromLocalTile<Coord::Row>(i_loc_tile) - i_begin;
-          auto& i2_tile = i2_tile_arr[to_sizet(i_subm_i2_arr)];
-
-          const SizeType nrows = std::min(dist.tileSize<Coord::Row>(i_gl_tile), n - i_gl_subm_el);
-          for (int i = 0; i < nrows; ++i) {
-            const SizeType ii = i2_tile({i, 0});
-            if (ii < k)
-              evec_tile({i, j_tile_el}) = delta_ptr[ii];
+
+  TileCollector tc{i_begin, i_end};
+
+  const SizeType n = problemSize(i_begin, i_end, dist);
+
+  const SizeType m_subm_el_lc = [=]() {
+    const auto i_loc_being = ij_begin_lc.row();
+    const auto i_loc_end = ij_begin_lc.row() + sz_loc_tiles.rows();
+    return dist.localElementDistanceFromLocalTile<Coord::Row>(i_loc_being, i_loc_end);
+  }();
+
+  const SizeType n_subm_el_lc = [=]() {
+    const auto i_loc_being = ij_begin_lc.col();
+    const auto i_loc_end = ij_begin_lc.col() + sz_loc_tiles.cols();
+    return dist.localElementDistanceFromLocalTile<Coord::Col>(i_loc_being, i_loc_end);
+  }();
+
+  auto bcast_evals = [i_begin, i_end,
+                      dist](common::Pipeline<comm::Communicator>& row_comm_chain,
+                            const std::vector<matrix::Tile<T, Device::CPU>>& eval_tiles) {
+    using dlaf::comm::internal::sendBcast_o;
+    using dlaf::comm::internal::recvBcast_o;
+
+    const comm::Index2D this_rank = dist.rankIndex();
+
+    std::vector<ex::unique_any_sender<>> comms;
+    comms.reserve(to_sizet(i_end - i_begin));
+
+    for (SizeType i = i_begin; i < i_end; ++i) {
+      const comm::IndexT_MPI evecs_tile_rank = dist.rankGlobalTile<Coord::Col>(i);
+      auto& tile = eval_tiles[to_sizet(i - i_begin)];
+
+      if (evecs_tile_rank == this_rank.col())
+        comms.emplace_back(ex::when_all(row_comm_chain(), ex::just(std::cref(tile))) |
+                           transformMPI(sendBcast_o));
+      else
+        comms.emplace_back(ex::when_all(row_comm_chain(), ex::just(evecs_tile_rank, std::cref(tile))) |
+                           transformMPI(recvBcast_o));
+    }
+
+    return ex::ensure_started(ex::when_all_vector(std::move(comms)));
+  };
+
+  auto all_reduce_in_place_o = [](const dlaf::comm::Communicator& comm, MPI_Op reduce_op,
+                                  const auto& data, MPI_Request* req) {
+    auto msg = comm::make_message(data);
+    DLAF_MPI_CHECK_ERROR(MPI_Iallreduce(MPI_IN_PLACE, msg.data(), msg.count(), msg.mpi_type(), reduce_op,
+                                        comm, req));
+  };
+
+  // Note: at least two column of tiles per-worker, in the range [1, getTridiagRank1NWorkers()]
+  const std::size_t nthreads = [size = sz_loc_tiles.cols()]() {
+    const std::size_t min_workers = 1;
+    const std::size_t available_workers = getTridiagRank1NWorkers();
+    const std::size_t ideal_workers = util::ceilDiv(to_sizet(size), to_sizet(2));
+    return std::clamp(ideal_workers, min_workers, available_workers);
+  }();
```

Comment:
Indeed, they were inside at some point, then I moved them out because I was thinking that probably some of them can be useful in general and can be added to the library itself. For the moment I wouldn't bother too much if you agree, because keeping them outside of the bulk reduces a bit the code cluttering in it (it is already too long and complex for my taste).